lib/matplotlib/legend.py

```diff
@@ -1337,6 +1337,12 @@ def _parse_legend_args(axs, *args, handles=None, labels=None, **kwargs):
         _api.warn_external("You have mixed positional and keyword arguments, "
                            "some input may be discarded.")
 
+    if (hasattr(handles, "__len__") and
```

Comment:
Is it better do do the `hasattr` checks or catch (and ignore) the `TypeError` fro just trying?

Reply:
I wondered the same. My reasoning was that since in the previous discussion it was agreed that only a warning should be raised , it made more sense to check for the attiribute rather than to do a try/except.
From a speed pov, I don't think there is going to be any big differences (try/except should be faster when __ len __ exists and viceversa). This [stack discussion ](https://stackoverflow.com/questions/903130/hasattr-vs-try-except-block-to-deal-with-non-existent-attributes)could also be useful.

With try/except would be something like:
```python
try:
    if len(handles) != len(labels):
        _api.warn_external(f"Mismatched number of handles and labels: "
                               f"len(handles) = {len(handles)} "
                               f"len(labels) = {len(labels)}")
except TypeError:
    pass
```
vs the current impl
https://github.com/matplotlib/matplotlib/blob/58a80fdbac119f1430383b0e610d546c11cdb891/lib/matplotlib/legend.py#L1340-L1345

I think using the hasattr() is cleaner, but I don't have a strong opinion about either.

In this file, hasattr is used once before:
https://github.com/matplotlib/matplotlib/blob/58a80fdbac119f1430383b0e610d546c11cdb891/lib/matplotlib/legend.py#L1251-L1257

All other cases use try/except.

Reply:
+0.1 on the version of the PR.
IMHO here: "explicit is better than implicit" > "it's better to ask for foregiveness than for permission", because TypeError is quite generic.

Reply:
I could see an argument for moving it into the `if handles and labels:` block, which is already the next check.

That _should_ make it so that both are at least iterable (escaping the one or both is None case), though I guess someone could pass e.g. `labels=(str(x) for x in itertools.count())`, which has no `len` and is relying on the `zip`, so still need _some_ guard against unbounded inputs.

Reply:
I think this is clearer. We're not obliged to jump extra hoops for user-friendly length checks if the user passes an iterable without a `len` implementation.